enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+     1. Admin reboots the device
+     1. Previous boot was healthy - back up the data
+     1. MicroShift cluster starts
+
+   - If the admin simply reboots the device without fixing problems
+     1. _Current boot's commit_ is the same as _previous boot's commit_
+        and backup exists, so it restores the backup
+     1. MicroShift cluster starts
+
+
+##### Commit pre-loaded to device is unhealthy, admin wants to stage a different one
+
+Device was pre-loaded with a commit that includes MicroShift.
+Because it is the only commit existing on the device, there's nothing to roll back to.
+To address this, admin wants to stage another commit with MicroShift.
+
+1. Pre-loaded commit is unhealthy
+1. greenboot doesn't perform reboots to heal the device
+1. Admin stops the MicroShift
+1. Admin removes MicroShift's data
+1. Admin stages another commit and reboots the device
+1. New commit starts
+1. There is no existing MicroShift's data, it is like first boot,
+   so MicroShift starts the cluster.
+
+##### Admin wants to manually roll back
+
+Both rollback and currently running commits are healthy,
+but admin wishes to roll back to the previous commit.
+
+1. Second commit is running and healthy
+1. Admin runs `rpm-ostree rollback` command or equivalent and reboot the device
+1. Device boots first commit
+1. _Previous boot_ was healthy so data is backed up
+1. _Current boot's commit_ is different from _previous boot's commit_
+   and backup for _current_ exists, so it is restored
+1. MicroShift starts the cluster
+
+##### Upgrade from 4.13 failed and system rolled back to 4.13
+
+Following scenario describes failed upgrade from 4.13
+and provides information on what should admin do to retry upgrade.
+
+1. Commit with MicroShift 4.13 is running
+1. Admin prepares and stages new commit that includes newer MicroShift,
+   and reboots the device.
+1. MicroShift's data exists, but metadata with version is missing,
+   so it is assumed that upgrade is from 4.13.
+   - Back up 4.13 data
+   - If upgrade from 4.13 is supported, then try to migrate the data.
+     Otherwise refuse (renders system unhealthy and causes a rollback).
+1. MicroShift starts the cluster
+1. System ends up unhealthy for any reason. It could be that:
+   - Upgrade was blocked or storage migration failed
+   - MicroShift was unhealthy
+   - Something else was unhealthy
+1. Device is rebooted by greenboot
+1. Because _previous boot_ was unhealthy,
+   there is no backup for _current boot's commit_,
+   _previous commit with MicroShift_ is the same as _rollback_:
+   restore backup of _previous commit with MicroShift_ and attempt to perform
+   the upgrade again.
+1. System happens to be consistently unhealthy and rolls back to commit with 4.13
+1. MicroShift cluster starts, we cannot predict if it fails or not
+1. Regardless if system is healthy or not, admin still wants to upgrade.
+1. Admin manually restores backup compatible with 4.13 and addresses everything
```

Comment:
why didn't the backup restoration happen automatically as part of the rollback to 4.13?

Reply:
Because we're targeting 4.14 for this enhancement and 4.13 doesn't have mechanisms for this.
MicroShift's data resides in `/var/` which is mutable part of filesystem and carried over across different deployments, so it's up to MicroShift to back it up or restore in right moments.

Reply:
ok, it might be worth an explicit discussion about why 4.13 is different from how this will work in 4.14+  (i know there is some discussion in various spots, but i don't think there's a cohesive summary of what 4.13 is "missing" and the implications of that)

Reply:
All of the logic for handling upgrades, backups, and rollbacks described in this enhancement will be missing from 4.13. We can make that more explicit in the text. We can also make clear that 4.13 is a preview release, and we are not required to support upgrading from 4.13.